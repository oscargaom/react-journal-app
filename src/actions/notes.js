import { db } from "../firebase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import { fileUpload } from "../helpers/fileUpload";

import Swal from "sweetalert2";

export const startNewNote = () => {
    /*  También podemos obtener el estado a través de una función que es enviada 
        por thunk como parámetro junto con el dispatch. Gracias a ello puedemos 
        tener acceso a todo el state y a los reducers y sería equivalente a usar
        el hook useSelector.
    */
    return async (dispatch, getState) => {
        // console.log(getState());
        const { auth: { uid } } = getState();
        // console.log(uid);

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});


export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        // console.log(notes);
        dispatch(setNotes(notes));
    };
};


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        // console.log('startSaveNote note');
        // console.log(note);
        const noteToUpdate = { ...note };
        // console.log(noteToUpdate);
        delete noteToUpdate.id;
        // console.log(noteToUpdate);

        if (!noteToUpdate.url) {
            delete noteToUpdate.url;
        }

        try {
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToUpdate);
        } catch (error) {
            Swal.fire('Cant saved your note', error.toString(), 'error')
        }

        // console.log('startSaveNote note.id');
        // console.log(note.id);
        dispatch(refreshNote(note.id, note));

        Swal.fire('Saved note', 'Your note has been saved', 'success');
    };
};

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    };
};

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: note
    };
};

export const startUploading = (file) => {
    return async (dispatch, getState) => {

        const { notes: { active } } = getState();

        // console.log(file);
        // console.log(active);

        Swal.fire({
            title: 'Uploading ...',
            text: 'Please wait ...',
            allowOutsideClick: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        // console.log(fileUrl);

        dispatch(startSaveNote(active));

        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { auth: { uid } } = getState();

        // console.log(`startDeleting uid: ${uid}`);

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})


export const cleanNotes = () => ({
    type: types.notesLogoutCleaning,
    payload: null
})