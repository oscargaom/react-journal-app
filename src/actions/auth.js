import { types } from "../types/types";
import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from "./ui";
import { cleanNotes } from "./notes";

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);
                dispatch(login(displayName, uid));
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // console.log(email);
        // console.log(password);
        // console.log(name);

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async (userCredential) => {
                // console.log(userCredential);
                const { user } = userCredential;
                await user.updateProfile({ displayName: name });
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);

                // console.log(user);
                // console.log(displayName);
                Swal.fire('Bienvenido', 'Tu cuenta ha sido creada', 'success');
                dispatch(login(uid, displayName));

            })
            .catch(error => {
                // console.log(error);
                Swal.fire('Register error', error.message, 'error')
            });
    };
}

/*  Esta es una acción asíncrona, por ello en vez de regresar un objeto regresa 
    un callback el cual el middleware thunk queda en espera de la ejecución de algún
    dispatch; en caso de que no se ejecutará algún dispatch simplmente termina. 
    El dispatch es proporcionado por el middleware thunk, así es que no debemos 
    preocuparnos por obtenerlo de algún otro lugar.
*/
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        // setTimeout(() => {
        //     dispatch(login('Gpe', '456789'));
        // }, 3000);


        dispatch(startLoading());

        // setTimeout(() => {
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then((userCredential) => {
                dispatch(finishLoading());
                const { user: { displayName, uid } } = userCredential;
                Swal.fire('Bienvenido',`Hola ${displayName}`,'success');
                dispatch(login(uid, displayName));
            })
            .catch(error => {
                // const {code} = error;
                dispatch(finishLoading());
                dispatch(login('', ''));
                Swal.fire('Login Error', error.message, 'error');
            });
        // }, 9000);
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());

        dispatch(cleanNotes());
    };
}



