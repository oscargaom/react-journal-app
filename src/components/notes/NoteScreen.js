import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useForm } from '../../hooks/useForm'
import { useSelector, useDispatch } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { notes: { active: note } } = useSelector(state => state);

    // console.log(note);
    const [formValues, formHandleInputChange, reset] = useForm(note);

    const { body, title, url , id} = note;
    
    /*  useRef almacena una variable mutable que en este caso es note.id la 
        cual no va a redibujar a todo el componente NoteScreen si esta cambia.
    */
    const noteId = useRef(note.id);
    // console.log(noteId);
    
    useEffect(() => {
        /*  Como useForm utiliza su propio state debemos modificar el valor de
            dicho estado para que sea redibujado el componente con los valores
            de la nota, ya que de lo contrario siempre muestra el primer valor.
        */
        if (noteId.current !== note.id) {
            reset(note);
            noteId.current = note.id;
        }
    }, [reset, note]);

    
    useEffect(() => {
        // console.log(formValues);
        dispatch(activeNote(formValues.id, {...formValues}));
    }, [formValues, dispatch])
    
    const handleDelete = () => {
        // console.log(id);
        dispatch(startDeleting(id));
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    name="title"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={formHandleInputChange}
                />

                <textarea
                    name="body"
                    placeholder="What happened today"
                    className="notes__textarea"
                    value={body}
                    onChange={formHandleInputChange}
                >
                </textarea>

                {
                    (url)
                    &&
                    (<div className="notes__image">
                        <img
                            src={url}
                            alt="imagen"
                        />
                    </div>)
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>
    )
}
