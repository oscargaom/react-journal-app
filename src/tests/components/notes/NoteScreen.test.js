import React from 'react'
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';


import { initialStateAuthNotesActiveNote } from '../../utils/initialState';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);


jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));

describe('Pruebas en el componente <NoteScreen />', () => {

    let store = mockStore(initialStateAuthNotesActiveNote);

    beforeEach(() => {
        store = mockStore(initialStateAuthNotesActiveNote);
        store.dispatch = jest.fn();
        jest.clearAllMocks();
    });

    test('debe mostrarse el snapshot correctamente', () => {

        const wrapper = mount(
            <Provider store={store}>
                <NoteScreen />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe ejecutar el dispatch para activeNote cuando se modifica el titulo de la nota ', () => {

        const wrapper = mount(
            <Provider store={store}>
                <NoteScreen />
            </Provider>
        );

        const texto = 'Pasear al perro';

        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: texto
            }
        }
        );

        /*  activeNote se ejecuta más de una vez porque la primera ocasión que lo hace es debido a la carga 
            inicla del componente y la segunda porque se dispara el useEffect debido a los cambios en formValues
            porque el cambio que se realizá en el title del input.
        */
        expect(activeNote).toHaveBeenCalledTimes(2);
        
        /*  Hacemos uso de toHaveBeenLastCalledWith para validar el último llamado, ya que el primero no nos 
            interesa saber con que valores fue llamado porque es el que se realiza con la carga por default.
        */
        expect(activeNote).toHaveBeenLastCalledWith(initialStateAuthNotesActiveNote.notes.active.id, {
            ...initialStateAuthNotesActiveNote.notes.active,
            title: texto,
        });
    });

})
