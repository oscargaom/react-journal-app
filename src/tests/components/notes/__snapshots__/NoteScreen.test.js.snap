// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <NoteScreen /> debe mostrarse el snapshot correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de agosto 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="MeditaciÃ³n"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="30 minutos."
    />
    <div
      className="notes__image"
    >
      <img
        alt="imagen"
        src="https://res.cloudinary.com/dhc6rsfox/image/upload/v1596500035/axoj3ozqkbfxxdch6tn1.jpg"
      />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
